examples/knative/powershell/kn-ps-vrni-databus/vrni-functions.ps1

```diff
@@ -0,0 +1,77 @@
+Import-Module CloudEvents.Sdk
+
+
+function New-vRNICloudEventsFromDatabus {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]$Data,
+        [Parameter(Mandatory = $true)]$cloudEventSource
+    )
+
+    # empty container for the cloud events. vRNI can send multiple events in a single request, so we need to split it up
+    $cloudEvents = @()
+    $jsonData = $Data | ConvertFrom-Json -AsHashtable
+
+    # see whether the payload body has problems or application updates and process them differently
+    # first up: problems
+    if ($jsonData.ContainsKey("EntityMessageList")) {
+        # loop through all problems and create CloudEvents from them to give back
+        # to the main function (which fires them into the event broker)
+        foreach ($problemEvent in $jsonData["EntityMessageList"]) {
+
+            if (!$problemEvent.ContainsKey("entity")) {
+                Write-host "$(Get-Date) - Wrong format for problem event, skipping:`n"
+                Write-host $problemEvent
+                continue
+            }
+
+            # use this as the event type
+            # the payload has a "status" field with "UPDATE", "OPEN", or "CLOSE" to figure out what the problem update is about
+            $cloudEventType = "com.vmware.event.router/vrni.problem.v0"
```

Comment:
Since the `problem` type can be different: UPDATE, OPEN and CLOSE, I think it would be nice to follow similiar pattern that vSphere uses to create unique types for each (e.g VM Created, VM Deleted, VM Updated)

You can easily do this with switch statement and the three types would be:
* com.vmware.event.router/vrni.problem.updated.v0
* com.vmware.event.router/vrni.problem.opened.v0
* com.vmware.event.router/vrni.problem.closed.v0

Note: The `status` field should be lower case and using past tenths

Reply:
That could work. I also thought about `/vrni.problem.$problem_name.v0` at one point, so it would be easier to subscribe to a single problem event type. There are currently 387 different events in vRNI.

But I couldn't figure out how to subscribe to a wildcard with the reacting function, nor find documentation about it. Do you know whether that's possible, or coming to Knative? If it is, this format may be even better? 

* com.vmware.event.router/vrni.problem.$problem_name.updated.v0
* com.vmware.event.router/vrni.problem.$problem_name.opened.v0
* com.vmware.event.router/vrni.problem.$problem_name.closed.v0

If the wildcard for events isn't there, I'll update it to include the status in the event. It makes sense, there are different actions possible based on if it's a new, updated, or closed event.

Btw, the `status` field comes from vRNI's payload, changing it to updated/opened/closed means manipulating the incoming data making it differ from the spec that vRNI has defined. Not sure if that's a good way to go?

Reply:
There's currently no wildcard support, its something folks have asked about. I know upstream Knative is looking into improving filtering, so hopefully that'll be possible in future. 

In terms of multiple event types, the downside of having a single event is that its not performant since you now need to add logic to your code to filter out on the thing you're interested in and ideally this is done outside of your function to keep it nice and simple. If there was only a handful, that could work but given there's over 380+ ... I'd say ideal experience is to have each type be its own event, this way someone can easily filter on event a, b, c and the function will never see events it doesn't wish to subscribe to. We have some vSphere Events like that, like an Alarm and its not great user experience for that reason

> changing it to updated/opened/closed means manipulating the incoming data making it differ from the spec that vRNI has defined. Not sure if that's a good way to go?

Well, you're not changing what the original status field means, you're simply using it to define your event type, which must be all lower case and pattern is typically past tenths as the event had already occurred. You can still maintain the status value which is in UPPER case but we just require the CloudEvent type to be defined based on guidelines. Let me know if that makes sense

Reply:
> In terms of multiple event types, the downside of having a single event is that its not performant since you now need to add logic to your code to filter out on the thing you're interested in and ideally this is done outside of your function to keep it nice and simple. If there was only a handful, that could work but given there's over 380+

At a higher view, there's only 1 event type for alerts: the problem. Considering there's no wildcard functionality yet, I'll keep it at the 3 types you suggested initially. So far, the majority of customers want to funnel all alerts into some other service (Slack/ServiceNow/etc), and defining 380+ VEBA functions to catch all alerts is not going to work. Only a few want to respond to a single alert.

> Well, you're not changing what the original status field means, you're simply using it to define your event type, which must be all lower case and pattern is typically past tenths as the event had already occurred.

I thought you meant the `status` field inside the body, but I see that you're referring to the status *inside* the `type` field now. ;-)

Updated the code to reflect.